storage/shard.go

```diff
@@ -1590,6 +1681,88 @@ func (s *dbShard) removeAnyFlushStatesTooEarly() {
+	return s.snapshotState.isSnapshotting, s.snapshotState.lastSuccessfulSnapshot
+}
+
+func (s *dbShard) markIsSnapshotting() {
+	s.snapshotState.Lock()
+	s.snapshotState.isSnapshotting = true
+	s.snapshotState.Unlock()
+}
+
+func (s *dbShard) markDoneSnapshotting(success bool, completionTime time.Time) {
+	s.snapshotState.Lock()
+	s.snapshotState.isSnapshotting = false
+	if success {
+		s.snapshotState.lastSuccessfulSnapshot = completionTime
+	}
+	s.snapshotState.Unlock()
+}
+
+// CleanupSnapshots examines the snapshot files for the shard that are on disk and
+// determines which can be safely deleted. A snapshot file is safe to delete if it
+// meets one of the following criteria:
+// 		1) It contains data for a block start that is out of retention (as determined
+// 		   by the earliestToRetain argument.)
+// 		2) It contains data for a block start that has already been successfully flushed.
+// 		3) It contains data for a block start that hasn't been flushed yet, but a more
+// 		   recent set of snapshot files (higher index) exists for the same block start.
+// 		   This is because snapshot files are cumulative, so once a new one has been
+//         written out it's safe to delete any previous ones for that block start.
+func (s *dbShard) CleanupSnapshots(earliestToRetain time.Time) error {
+	filePathPrefix := s.opts.CommitLogOptions().FilesystemOptions().FilePathPrefix()
+	multiErr := xerrors.NewMultiError()
+	snapshotFiles, err := s.snapshotFilesFn(filePathPrefix, s.namespace.ID(), s.ID())
+	if err != nil {
+		return err
+	}
+
+	sort.Slice(snapshotFiles, func(i, j int) bool {
+		// Make sure they're sorted by blockStart/Index in ascending order.
+		if snapshotFiles[i].BlockStart.Equal(snapshotFiles[j].BlockStart) {
+			return snapshotFiles[i].Index < snapshotFiles[j].Index
+		}
+		return snapshotFiles[i].BlockStart.Before(snapshotFiles[j].BlockStart)
+	})
+
+	filesToDelete := []string{}
+
+	for i := 0; i < len(snapshotFiles); i++ {
+		curr := snapshotFiles[i]
+
+		if curr.BlockStart.Before(earliestToRetain) {
+			// Delete snapshot files for blocks that have fallen out
+			// of retention.
+			filesToDelete = append(filesToDelete, curr.Files...)
+			continue
+		}
+
+		if s.FlushState(curr.BlockStart).Status == fileOpSuccess {
+			// Delete snapshot files for any block starts that have been
+			// successfully flushed.
+			filesToDelete = append(filesToDelete, curr.Files...)
+			continue
+		}
+
+		if i+1 < len(snapshotFiles) &&
+			snapshotFiles[i+1].BlockStart == curr.BlockStart &&
+			snapshotFiles[i+1].Index > curr.Index &&
+			snapshotFiles[i+1].HasCheckpointFile() {
+			// Delete any snapshot files which are not the most recent
+			// for that block start, but only of the set of snapshot files
+			// with the higher index is complete (checkpoint file exists)
+			filesToDelete = append(filesToDelete, curr.Files...)
+			continue
+		}
+	}
+
+	err = s.deleteFilesFn(filesToDelete)
+	return multiErr.Add(err).FinalError()
```

Comment:
i don't think you need to use multiErr at all in this function. can simplify this to just `return s.deleteFilesFn(..)`

Reply:
you're right, I don't